### PR TITLE
Add feature groups for linting

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

This implements the same optimizations outlined in #877, which have been applied in #892 for tests. We forgot Clippy in the initial PR, which has been applied in this PR after being pointed out by https://github.com/hashintel/hash/pull/794#discussion_r949698353.

## 🔗 Related links

* Discussion about time complexity and motivation: #877 
* Initial PR: #892 

## 🔍 What does this change?

`cargo make clippy` will only run for a fraction of feature sets on PRs, but with all permutations on release.

## 📜 Does this require a change to the docs?

Not user-facing, just CI.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -209,6 +209,11 @@ jobs:
       - name: Run tests
         working-directory: ${{ matrix.directory }}
         run: cargo +${{ matrix.toolchain }} make --profile ${{ matrix.profile }} test --no-fail-fast
+      
+      - name: Run lints
+        working-directory: ${{ matrix.directory }}
+        run: cargo +${{ matrix.toolchain }} make --profile ${{ matrix.profile }} clippy -- -D warnings
+
```
Comment (TimDiekmann) [inline .github/workflows/rust.yml:216]: Let's put this above the tests and mark the tests as lints are usually faster

File: .github/scripts/rust/Makefile.toml [hunk 2]
```diff
@@ -12,6 +12,10 @@ CARGO_RUN_FLAGS = ""
 CARGO_FORMAT_FLAGS = ""
 CARGO_FORMAT_HACK_FLAGS = "--workspace"
 CARGO_CLIPPY_FLAGS = "--no-deps --tests"
+# This needs to be here, as error-stack uses it, and _BASE **needs** to be defined before
+# the variable using it has been defined. This is done per suggestion of
+# https://github.com/sagiegurari/cargo-make/issues/685
+CARGO_CLIPPY_HACK_FLAGS_BASE = ""
```
Comment (TimDiekmann) [inline .github/scripts/rust/Makefile.toml:18]: This is annoying but probably the easiest way for us 🙁

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Thanks for picking this up! 🎉

File: .github/scripts/rust/Makefile.toml (hunk 2 above)
Comment (indietyp) [reply .github/scripts/rust/Makefile.toml:18]: We will be able to remove this once https://github.com/sagiegurari/cargo-make/pull/688 is merged.

Comment (TimDiekmann) [reply .github/scripts/rust/Makefile.toml:18]: Can't wait for it 😆

File: packages/libs/error-stack/Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "error-stack"
-version = "0.1.1"
+version = "0.2.0"
```
Comment (TimDiekmann) [inline packages/libs/error-stack/Cargo.toml:3]: The `publish` job works as expected 🎉 
```suggestion
version = "0.1.1"
```

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**